### ISSUE TITLE
use Rucio to resolve LFN2PFN

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#6102]
On 20/07/2020 08:47, Eric W Vaandering wrote:
> I think you can just use this: https://rucio.readthedocs.io/en/latest/api/rse.html#rucio.client.rseclient.RSEClient.lfns2pfns
> 
> To whit: "The PFNs are generated for the RSE /regardless/ of whether a replica exists for the LFN.”
> 
> If that’s not enough, then let's talk and we’ll figure something out.
> 
>

**Comment by dciangot** [comment]
Main items here:
- inclusion of the apis into wmcore
- at least T3s (temp + normal) and T1 Temp RSEs have to be carefully checked

**Comment by dciangot** [comment]
Hi @amaltaro, I can start to look into this, also with some effort to make changes where needed if you like. Should we have a quick chat on how to implement this, or do you have already a plan in mind?

**Comment by amaltaro** [comment]
@dciangot Hi Diego,

the WMCore Rucio wrapper already has the skeleton for that functionality:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/Rucio/Rucio.py#L273

so you'd need to implement the correct rucio-clients call in there and any other logic required by CRAB.

Feel free to also extend/update that specific unit test in here:
https://github.com/dmwm/WMCore/blob/master/test/python/WMCore_t/Services_t/Rucio_t/Rucio_t.py#L224

and of course, let us know in case you need any other help here.

**Comment by dciangot** [comment]
Perfect, do you need an issue on WMCore repo too? Or once I'll have a prototype we can discuss it on the PR?

**Comment by amaltaro** [comment]
If you could create a WMCore GH issue (new feature), that would be great! Don't worry about expanding all the requirements on the issue right now, we can always get back to them and update the issue in the future.

**Comment by belforte** [comment]
overall plan, from @dciangot 

- The draft PR with all the places that needs a change is here:
https://github.com/dmwm/CRABServer/pull/6183

- The  only place to  implement a cache are transfer scripts

- other things? 

Other components involved:
- `src/python/TaskWorker/Actions/DagmanCreator.py`
- `src/python/CRABInterface/HTCondorDataWorkflow.py`
- `src/python/CRABInterface/DataWorkflow.py`

**Comment by belforte** [comment]
PR #6183 touches these files, I am surprised that REST has reasons to call lfn2pfn, so am taking this chance to review if instead of moving from PhEDEx to Rucio there, we can simply drop code. And avoid being in a hurry to deploy a fixed REST before PhEDEx is shut off.

1. scripts/dag_bootstrap.sh
1. scripts/task_process/FTS_Transfers.py
1. scripts/task_process/task_proc_wrapper.sh
1. src/python/CRABInterface/DataWorkflow.py
1. src/python/CRABInterface/HTCondorDataWorkflow.py
1. src/python/CRABInterface/Utils.py
1. src/python/TaskWorker/Actions/DagmanCreator.py
1. src/python/TransferInterface/RegisterFiles.py 

It is OK to change things in `task_process` and 'TransferInterface`, but then ?

**Comment by dciangot** [comment]
Dagman creator? I suspect we should keep it too, no?

**Comment by belforte** [comment]
1. `scripts/dag_bootstrap.sh` is about adding Rucio to `$PYTHONPATH`.  So it s OK
1. `src/python/CRABInterface/DataWorkflow.py` 
I will review them all :-)
> Dagman creator? I suspect we should keep it too, no?

**Comment by belforte** [comment]
1. `scripts/dag_bootstrap.sh` is about adding Rucio to `$PYTHONPATH`.  So it s OK
1. `src/python/CRABInterface/DataWorkflow.py`  is only preparatory for using Rucio in `HTCondorDataWorkflow`
1 . `src/python/CRABInterface/HTCondorDataWorkflow.py`

**Comment by belforte** [comment]
@dciangot I think you overlooked this one, even if I do not know where it is used:
https://github.com/dmwm/CRABServer/blob/7334ec700a5dce0e69fb6b27efd01e0c467d7f38/src/python/ServerUtilities.py#L301-L316

**Comment by belforte** [comment]
I have prepared a PR which removes all PhEDEx calls from REST https://github.com/dmwm/CRABServer/pull/6211
@dciangot please remove changes to those files from your PR  https://github.com/dmwm/CRABServer/pull/6183
to avoid conflict and restoring of useless code

**Comment by belforte** [comment]
the new `rucioclient.getPFN` is now available in wmcore 1.3.6.crab4
https://github.com/dmwm/WMCore/blob/086b2878cdda2eaa16646a8d20029fdfafe8a963/src/python/WMCore/Services/Rucio/Rucio.py#L260-L274

**Comment by belforte** [comment]
Update:
* CRABServer now makes no PhEDEx and no Rucio calls. Tested in preprod. WIll go in November CMSWEB update
* TaskWorker components will be PhEDEx free and moved to new `rucioclient.getPFN` with #6221 
* only these needs to be changed (from list in  https://github.com/dmwm/CRABServer/issues/6102#issuecomment-712895512 ) and I plan to reuse as much as possible of Diego's code from #6183 hopefully the new, simpler `getPfn` will allow for some code simplification. But it will be cleaner to make a new PR
* once this is done and working, will worry about caching. But as long as there's one call to Rucio at every ieration of scripts/components which only run every few minutes apart, it is not urgent.

**Comment by belforte** [comment]
solved via https://github.com/dmwm/CRABServer/pull/6222
